mixer/adapter/kubernetesenv/kubernetesenv_test.go

```diff
@@ -384,6 +392,83 @@ func TestKubegen_Generate(t *testing.T) {
 	}
 }
 
+func createMultiClusterSecret(k8s *fake.Clientset) error {
+	data := map[string][]byte{}
+	secret := v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      testSecretName,
+			Namespace: testSecretNameSpace,
+			Labels: map[string]string{
+				"istio/multiCluster": "true",
+			},
+		},
+		Data: map[string][]byte{},
+	}
+
+	data["testRemoteCluster"] = []byte("Test")
+	secret.Data = data
+	_, err := k8s.CoreV1().Secrets(testSecretNameSpace).Create(&secret)
+	return err
+}
+
+func deleteMultiClusterSecret(k8s *fake.Clientset) error {
+	var immediate int64
+
+	return k8s.CoreV1().Secrets(testSecretNameSpace).Delete(
+		testSecretName, &metav1.DeleteOptions{GracePeriodSeconds: &immediate})
+}
+
+func mockLoadKubeConfig(kubeconfig []byte) (*clientcmdapi.Config, error) {
+	return &clientcmdapi.Config{}, nil
+}
+
+func mockCreateInterfaceFromClusterConfig(clusterConfig *clientcmdapi.Config) (kubernetes.Interface, error) {
+	return fake.NewSimpleClientset(), nil
+}
+
+func Test_KubeSecretController(t *testing.T) {
+	secretcontroller.LoadKubeConfig = mockLoadKubeConfig
+	secretcontroller.CreateInterfaceFromClusterConfig = mockCreateInterfaceFromClusterConfig
+
+	clientset := fake.NewSimpleClientset()
+	b := newBuilder(func(string, adapter.Env) (kubernetes.Interface, error) {
+		return clientset, nil
+	})
+
+	// Call kube Build function which will start the secret controller.
+	// Sleep to allow secret process to start.
+	_, err := b.Build(context.Background(), test.NewEnv(t))
+	if err != nil {
+		t.Fatalf("error building adapter: %v", err)
+	}
+	time.Sleep(10 * time.Millisecond)
+
+	// Create the multicluster secret. Sleep to allow created remote
+	// controller to start and callback add function to be called.
+	err = createMultiClusterSecret(clientset)
+	if err != nil {
+		t.Fatalf("Unexpected error on secret create: %v", err)
+	}
+	time.Sleep(200 * time.Millisecond)
+
+	// Test
+	if len(b.controllers) != 2 {
```

Comment:
continuing: could we maybe use a Ticker or something here and timeout with a fail after 5 tries or whatever while waiting on the second controller?

Reply:
Still uses sleeps but more resilient. pseudo-code obviously. 
MAX_TIMEOUTS = 30 // allow for ~double what i see when testing
for i = 0; i < MAX_TIMEOUTS; i++ {
    if len(b.controllers) == 2 {
          break
    } else {
          sleep(10 * milli)
    }
}
if i >= MAX_TIMEOUTS {
     fatal(error msg)
}

Could use this type of checking below also.

Thoughts?